File: internal/event/event.go
// Package event defines the normalized calendar event model shared by the
// Radarr/Sonarr clients, the iCal generator, and the JSON API.
package event

import (
	"fmt"
	"sort"
	"time"
)

type Kind string

const (
	KindEpisode       Kind = "episode"
	KindMovieCinema   Kind = "movie-cinema"
	KindMovieDigital  Kind = "movie-digital"
	KindMoviePhysical Kind = "movie-physical"
)

type Event struct {
	UID        string    `json:"uid"`
	Instance   string    `json:"instance"`
	Source     string    `json:"source"` // "radarr" | "sonarr"
	Kind       Kind      `json:"kind"`
	Title      string    `json:"title"`    // movie title or series title
	Subtitle   string    `json:"subtitle"` // episode title, "" for movies
	Season     int       `json:"season"`
	Episode    int       `json:"episode"`
	Start      time.Time `json:"start"`
	End        time.Time `json:"end"`
	AllDay     bool      `json:"allDay"`
	Downloaded bool      `json:"downloaded"`
	Overview   string    `json:"overview"`
	PosterURL  string    `json:"posterUrl"`
}

// EpisodeUID returns the stable UID for a Sonarr episode event. UIDs must
// never change for the same upstream entity or subscribed calendar clients
// will duplicate events on refresh.
func EpisodeUID(instance string, episodeID int64) string {
	return fmt.Sprintf("sonarr-%s-%d@arr-cal-proxy", instance, episodeID)
}

// MovieUID returns the stable UID for one release-date event of a Radarr
// movie. release is one of "cinema", "digital", "physical".
func MovieUID(instance string, movieID int64, release string) string {
	return fmt.Sprintf("radarr-%s-%d-%s@arr-cal-proxy", instance, movieID, release)
}

// SxxEyy formats season/episode as "S02E05".
func (e Event) SxxEyy() string {
	return fmt.Sprintf("S%02dE%02d", e.Season, e.Episode)
}

// Sort orders events by start time, then UID for determinism.
func Sort(events []Event) {
	sort.Slice(events, func(i, j int) bool {
		if !events[i].Start.Equal(events[j].Start) {
			return events[i].Start.Before(events[j].Start)
		}
		return events[i].UID < events[j].UID
	})
}

File: src/domain/event.ts
import type { EventDto, Kind, Source } from '../../shared/api.ts'

export interface CalendarEvent {
  uid: string
  instance: string
  source: Source
  kind: Kind
  title: string
  subtitle: string
  season: number
  episode: number
  start: Date
  end: Date
  allDay: boolean
  downloaded: boolean
  overview: string
  posterUrl: string
  providerIds?: Record<string, string>
  jellyfinUrl?: string
}

export function episodeUid(instance: string, episodeId: number): string {
  return `sonarr-${instance}-${episodeId}@calthing`
}

export function movieUid(
  instance: string,
  movieId: number,
  release: 'cinema' | 'digital' | 'physical',
): string {
  return `radarr-${instance}-${movieId}-${release}@calthing`
}

export function sxxEyy(event: Pick<CalendarEvent, 'season' | 'episode'>): string {
  return `S${String(event.season).padStart(2, '0')}E${String(event.episode).padStart(2, '0')}`
}

export function sortEvents(events: CalendarEvent[]): void {
  events.sort((a, b) => {
    const byStart = a.start.getTime() - b.start.getTime()
    if (byStart !== 0) return byStart
    return a.uid < b.uid ? -1 : a.uid > b.uid ? 1 : 0
  })
}

export function utcStartOfDay(date: Date): Date {
  return new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate()))
}

export function addUtcDays(date: Date, days: number): Date {
  const result = new Date(date)
  result.setUTCDate(result.getUTCDate() + days)
  return result
}

function toRfc3339(date: Date): string {
  return date.toISOString().replace(/\.000Z$/, 'Z').replace(/(\.\d*?[1-9])0+Z$/, '$1Z')
}

export function toEventDto(event: CalendarEvent): EventDto {
  return {
    uid: event.uid,
    instance: event.instance,
    source: event.source,
    kind: event.kind,
    title: event.title,
    subtitle: event.subtitle,
    season: event.season,
    episode: event.episode,
    start: toRfc3339(event.start),
    end: toRfc3339(event.end),
    allDay: event.allDay,
    downloaded: event.downloaded,
    overview: event.overview,
    posterUrl: event.posterUrl,
    jellyfinUrl: event.jellyfinUrl ?? '',
  }
}
